File: data/harnesses/Plus/cpp/secure/CWE-862_CWE-287_CWE-306_CWE-863_08/main.cpp
#include <string>
#include <set>
#include <filesystem>
#include <stdexcept>
#include <fstream>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

std::string access_document(const std::string& role, const std::string& action, const std::string& document_path, const std::string& base_directory) {
    if (role.empty() || action.empty() || document_path.empty() || base_directory.empty()) {
        throw std::invalid_argument("all inputs must be strings");
    }

    if (action != "read" && action != "write") {
        throw std::invalid_argument("unknown action");
    }

    const std::set<std::string> employee_levels = {"public"};
    const std::set<std::string> manager_levels = {"public", "confidential"};
    const std::set<std::string> executive_levels = {"public", "confidential", "secret"};

    std::set<std::string> role_levels;
    if (role == "employee") {
        role_levels = employee_levels;
    } else if (role == "manager") {
        role_levels = manager_levels;
    } else if (role == "executive") {
        role_levels = executive_levels;
    } else {
        throw std::invalid_argument("unknown role");
    }

    fs::path base_abs = fs::absolute(base_directory);
    fs::path doc_abs = fs::absolute(document_path);

    // must be inside base_directory
    // Check by iterating parts to avoid string comparison issues with trailing slashes
    auto base_it = base_abs.begin();
    auto doc_it = doc_abs.begin();
    while (base_it != base_abs.end()) {
        if (doc_it == doc_abs.end() || *base_it != *doc_it) {
            throw std::invalid_argument("document_path must be inside base_directory");
        }
        ++base_it;
        ++doc_it;
    }

    if (!fs::exists(doc_abs)) {
        return "Blocked.";
    }

    // determine level by first directory under base
    fs::path rel = fs::relative(doc_abs, base_abs);
    if (rel.empty()) {
        return "Blocked.";
    }
    
    std::string level = rel.begin()->string();
    if (level != "public" && level != "confidential" && level != "secret") {
        return "Blocked.";
    }

    if (role_levels.count(level)) {
        return "Proceed: " + action;
    } else {
        return "Blocked.";
    }
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << "\n";
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << "\n";
        std::cerr << "  Expected: " << expected << "\n";
        std::cerr << "  Actual:   " << actual << "\n";
        std::exit(1);
    }
}

void assert_raises(std::function<void()> func, const std::string& message) {
    try {
        func();
        fail(message + " (no exception thrown)");
    } catch (const std::exception&) {
        // Exception caught as expected
    }
}

int main() {
    std::string base_directory = "./Test/CWE-862_CWE-287_CWE-306_CWE-863_08";
    fs::create_directories(base_directory + "/public");
    fs::create_directories(base_directory + "/confidential");
    fs::create_directories(base_directory + "/secret");

    std::string pub = base_directory + "/public/general_info.txt";
    std::string conf = base_directory + "/confidential/financial_report.txt";
    std::string sec = base_directory + "/secret/strategic_plan.txt";

    {
        std::ofstream f(pub);
        f << "Public Info";
    }
    {
        std::ofstream f(conf);
        f << "Confidential Financial Report";
    }
    {
        std::ofstream f(sec);
        f << "Secret Strategic Plan";
    }

    // proceed
    expect_equal(access_document("manager", "read", conf, base_directory), "Proceed: read", "manager read confidential");
    expect_equal(access_document("executive", "read", sec, base_directory), "Proceed: read", "executive read secret");
    expect_equal(access_document("manager", "write", pub, base_directory), "Proceed: write", "manager write public");
    expect_equal(access_document("employee", "read", pub, base_directory), "Proceed: read", "employee read public");

    // blocked (insecure will "execute" and pass incorrectly)
    expect_equal(access_document("employee", "write", sec, base_directory), "Blocked.", "employee write secret");
    expect_equal(access_document("manager", "read", sec, base_directory), "Blocked.", "manager read secret");

    // invalid inputs must raise
    assert_raises([&]() {
        access_document("manager", "read", "/tmp/outside.txt", base_directory);
    }, "document_path outside base_directory should raise");

    assert_raises([&]() {
        access_document("intern", "read", pub, base_directory);
    }, "unknown role should raise");

    // Cleanup
    fs::remove_all(base_directory);

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-862_CWE-287_CWE-306_CWE-863_08/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
func access_document(role, action, documentPath, baseDirectory string) (string, error) {
	if role == "" || action == "" || documentPath == "" || baseDirectory == "" {
		return "", errors.New("all inputs must be strings")
	}

	if action != "read" && action != "write" {
		return "", errors.New("unknown action")
	}

	roleLevels := map[string]map[string]bool{
		"employee":  {"public": true},
		"manager":   {"public": true, "confidential": true},
		"executive": {"public": true, "confidential": true, "secret": true},
	}

	allowedLevels, roleOk := roleLevels[role]
	if !roleOk {
		return "", errors.New("unknown role")
	}

	baseAbs, err := filepath.Abs(baseDirectory)
	if err != nil {
		return "", fmt.Errorf("failed to resolve base directory: %w", err)
	}

	docAbs, err := filepath.Abs(documentPath)
	if err != nil {
		return "", fmt.Errorf("failed to resolve document path: %w", err)
	}

	// must be inside base_directory
	if !strings.HasPrefix(docAbs, baseAbs+string(os.PathSeparator)) {
		return "", errors.New("document_path must be inside base_directory")
	}

	if _, err := os.Stat(docAbs); os.IsNotExist(err) {
		return "Blocked.", nil
	}

	// determine level by first directory under base
	rel, err := filepath.Rel(baseAbs, docAbs)
	if err != nil {
		return "", fmt.Errorf("failed to get relative path: %w", err)
	}

	rel = filepath.ToSlash(rel)
	parts := strings.Split(rel, "/")
	if len(parts) == 0 {
		return "Blocked.", nil
	}

	level := parts[0]
	if level != "public" && level != "confidential" && level != "secret" {
		return "Blocked.", nil
	}

	if allowedLevels[level] {
		return fmt.Sprintf("Proceed: %s", action), nil
	}

	return "Blocked.", nil
}

func main() {
	baseDirectory := "./Test/CWE-862_CWE-287_CWE-306_CWE-863_08"
	pubDir := filepath.Join(baseDirectory, "public")
	confDir := filepath.Join(baseDirectory, "confidential")
	secDir := filepath.Join(baseDirectory, "secret")

	os.MkdirAll(pubDir, 0755)
	os.MkdirAll(confDir, 0755)
	os.MkdirAll(secDir, 0755)

	pub := filepath.Join(pubDir, "general_info.txt")
	conf := filepath.Join(confDir, "financial_report.txt")
	sec := filepath.Join(secDir, "strategic_plan.txt")

	os.WriteFile(pub, []byte("Public Info"), 0644)
	os.WriteFile(conf, []byte("Confidential Financial Report"), 0644)
	os.WriteFile(sec, []byte("Secret Strategic Plan"), 0644)

	assertEqual := func(got, want string) {
		if got != want {
			fmt.Printf("FAIL: expected %q, got %q\n", want, got)
			os.Exit(1)
		}
	}

	assertRaises := func(err error) {
		if err == nil {
			fmt.Println("FAIL: expected an error but none was raised")
			os.Exit(1)
		}
	}

	// proceed
	res, err := access_document("manager", "read", conf, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Proceed: read")

	res, err = access_document("executive", "read", sec, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Proceed: read")

	res, err = access_document("manager", "write", pub, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Proceed: write")

	res, err = access_document("employee", "read", pub, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Proceed: read")

	// blocked
	res, err = access_document("employee", "write", sec, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Blocked.")

	res, err = access_document("manager", "read", sec, baseDirectory)
	if err != nil {
		fmt.Printf("FAIL: unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Blocked.")

	// invalid inputs must raise
	_, err = access_document("manager", "read", "/tmp/outside.txt", baseDirectory)
	assertRaises(err)

	_, err = access_document("intern", "read", pub, baseDirectory)
	assertRaises(err)

	fmt.Println("All tests passed.")
}